### PR TITLE
Spi migrate users ups into test spi

### PR DESCRIPTION
**Pull request by AMEIJER1**
## Links to Jira Tickets

- {Include a link to all applicable Jira tickets}

## Description of Changes

- {List all relevant code changes. Include any changes to the business workflow that might not be obvious to the reviewers of this PR.}

## Testing Notes

- {List any relevant testing considerations, necessary pre-reqs, and areas of the app to focus on. Specifically, include anything that will help the reviewers of this PR verify the code is functioning as expected.}

### CODE REVIEW

File: spi/src/index.ts [hunk 1]
```diff
@@ -49,6 +49,9 @@ async function main() {
     // STEP 6. Transforms SPI Design Components into SIMS Sampling Sites
     await transformSampleSites(connection);
```
Comment (mauberti-bc) [inline spi/src/index.ts:59]: It looks like the same function is being called twice

File: spi/src/transformations/project.ts [hunk 2]
```diff
@@ -29,24 +29,65 @@ export const transformProjects = async (connection: IDBConnection): Promise<void
...
+
+        CASE 
+        WHEN p.coordinator IS NOT NULL AND w_mapping.full_name LIKE '%' || TRIM(BOTH ', ' FROM p.coordinator) || '%' AND spp.email_address LIKE '%@gov.bc.ca%' THEN 
+            (SELECT project_role_id FROM biohub.project_role WHERE name = 'Coordinator')
+        WHEN NOT EXISTS (
+            SELECT 1 
+            FROM biohub.project_participation 
+            WHERE project_id = w_mapping.project_id 
+              AND project_role_id = (SELECT project_role_id FROM biohub.project_role WHERE name = 'Coordinator')
+        ) THEN (SELECT project_role_id FROM biohub.project_role WHERE name = 'Coordinator'), (SELECT system_user_id FROM biohub.system_user WHERE user_identifier = 'spi')
```
Comment (mauberti-bc) [inline spi/src/transformations/project.ts]: Is there a syntax error here where the SPI user is being selected?

File: spi/src/transformations/project.ts [hunk 3]
```diff
@@ -29,24 +29,65 @@ export const transformProjects = async (connection: IDBConnection): Promise<void
...
+         FROM public.migrate_spi_user_deduplication 
+         WHERE w_mapping.person_id = ANY (spi_person_ids)
+        ), 
+        (SELECT system_user_id 
+         FROM biohub.system_user 
+         WHERE user_identifier = 'spi')
+        ) AS system_user_id,
+
+        CASE 
+        WHEN p.coordinator IS NOT NULL AND w_mapping.full_name LIKE '%' || TRIM(BOTH ', ' FROM p.coordinator) || '%' AND spp.email_address LIKE '%@gov.bc.ca%' THEN
```
Comment (mauberti-bc) [inline spi/src/transformations/project.ts]: The case when can be simplified

```
CASE
  WHEN TRIM(p.coordinator) = w_mapping.full_name AND spp.email_address LIKE '%@gov.bc.ca%' THEN 
    (SELECT project_role_id FROM biohub.project_role WHERE name = 'Coordinator')
  WHEN spp.email_address LIKE '%@gov.bc.ca%' THEN 
    (SELECT project_role_id FROM biohub.project_role WHERE name = 'Collaborator')
  ELSE 
    (SELECT project_role_id FROM biohub.project_role WHERE name = 'Observer')
END
```

File: spi/src/transformations/study-area.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    ),
+    buffered_geographies AS (
+        SELECT 
+            sp.survey_id,
+            sp.study_area_name,
+            sp.study_area_description,
+            sp.convex_hull,
+            sp.geo,
+            sp.so_count,
+            (0.2 * sqrt(ST_Area(sp.convex_hull)) + 0.05 * pow((ST_Area(sp.convex_hull) / sp.so_count), 0.75)) AS buffer_value
```
Comment (mauberti-bc) [inline spi/src/transformations/study-area.ts]: We should add a comment in the code indicating that these constants came from the Python script for inferring study areas based on observations.

File: spi/src/transformations/user.ts [hunk 5]
```diff
@@ -29,39 +29,54 @@ export const transformUsers = async (connection: IDBConnection): Promise<void> =
...
+        'Migrated from SPI as user ' || md.id,
+        COALESCE(spp.email_address, 'default') AS email
     FROM 
-      migrate_spi_user_deduplication;
+        migrate_spi_user_deduplication md
+    LEFT JOIN 
+        public.spi_secure_persons spp ON spp.first_name = md.given_name AND spp.last_name = md.family_name
+    LEFT JOIN 
+        w_existing_users eu ON eu.given_name = md.given_name AND eu.family_name = md.family_name
+    WHERE eu.system_user_id IS NULL;
```
Comment (mauberti-bc) [inline spi/src/transformations/user.ts:64]: eu.system_user_id will never be null because it is the primary key

Comment (AMEIJER1) [reply spi/src/transformations/user.ts:64]: We are left-joining eu onto deduplication table (which is the from table in the insert statement) -- in this case, everything from deduplication is joined to only matching records in eu, wouldnt therefore any deduplication records without a match in eu have a null eu.system_user_id as they haven't been "inserted" yet?

File: spi/src/index.ts (hunk 1 above)
Comment (AMEIJER1) [reply spi/src/index.ts:59]: fixed! and added other functions from the remaining completed transformations in

File: spi/src/transformations/project.ts (hunk 3 above)
Comment (AMEIJER1) [reply spi/src/transformations/project.ts]: simplified in code! Would this still ensure that when no coordinator is set, that the spi user is set as coordinator?

File: spi/src/transformations/study-area.ts (hunk 4 above)
Comment (AMEIJER1) [reply spi/src/transformations/study-area.ts]: comment inserted!

**Pull request merged**